File: node/src/ecmr.ts
import { HttpClient } from './http-client';
import type { ShipzyConfig } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface EcmrListItem {
    id: string;
    document_no: string;
    status: string;
    created_at: string;
}

export interface EcmrListResponse {
    data: EcmrListItem[];
    total: number;
    page: number;
    page_size: number;
}

// ============ Client ============

export class EcmrClient extends HttpClient {
    async list(params: { page?: number; pageSize?: number } = {}): Promise<ApiResult<EcmrListResponse>> {
        const q = this.buildQuery({ page: params.page, page_size: params.pageSize });
        return this.request(`/api/v1/shipment/ecmr/list${q}`);
    }

    async get(id: string): Promise<ApiResult<Record<string, any>>> {
        return this.request(`/api/v1/shipment/ecmr/${id}`);
    }

    async create(data: Record<string, unknown>): Promise<ApiResult<Record<string, any>>> {
        return this.request('/api/v1/shipment/ecmr/create', 'POST', data);
    }

    async generateFromOrder(orderId: string): Promise<ApiResult<Record<string, any>>> {
        return this.request('/api/v1/shipment/ecmr/generate-from-order', 'POST', { order_id: orderId });
    }

    async sign(id: string): Promise<ApiResult<Record<string, any>>> {
        return this.request(`/api/v1/shipment/ecmr/${id}/sign`, 'POST', {});
    }

    async update(id: string, data: Record<string, unknown>): Promise<ApiResult<Record<string, any>>> {
        return this.request(`/api/v1/shipment/ecmr/${id}/update`, 'POST', data);
    }

    async cancel(id: string): Promise<ApiResult<Record<string, any>>> {
        return this.request(`/api/v1/shipment/ecmr/${id}/cancel`, 'POST', {});
    }

    async validate(id: string): Promise<ApiResult<Record<string, any>>> {
        return this.request(`/api/v1/shipment/ecmr/${id}/validate`, 'POST', {});
    }

    async submitToAuthority(id: string): Promise<ApiResult<Record<string, any>>> {
        return this.request(`/api/v1/shipment/ecmr/${id}/submit-to-authority`, 'POST', {});
    }

    async pdf(id: string): Promise<ApiResult<{ status: string; pdf_url?: string }>> {
        return this.request(`/api/v1/shipment/ecmr/${id}/pdf`, 'POST', {});
    }
}

File: rust/src/ecmr.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;
use crate::types::{EcmrListItem, PaginatedResponse, PdfResponse};

pub struct EcmrClient {
    pub(crate) inner: HttpClient,
}

impl EcmrClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }
}

impl EcmrClient {
    pub async fn list(
        &self,
        page: i32,
        page_size: i32,
    ) -> Result<PaginatedResponse<EcmrListItem>> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/list?page={}&page_size={}", page, page_size),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn get(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}", id),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<serde_json::Value> {
        self.inner
            .request(
                "/api/v1/shipment/ecmr/create",
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn generate_from_order(
        &self,
        order_id: &str,
    ) -> Result<serde_json::Value> {
        self.inner
            .request(
                "/api/v1/shipment/ecmr/generate-from-order",
                Method::POST,
                Some(serde_json::json!({ "order_id": order_id })),
            )
            .await
    }

    pub async fn update(&self, id: &str, body: serde_json::Value) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}/update", id),
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn cancel(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}/cancel", id),
                Method::POST,
                None,
            )
            .await
    }

    pub async fn validate(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}/validate", id),
                Method::POST,
                None,
            )
            .await
    }

    pub async fn submit_to_authority(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}/submit-to-authority", id),
                Method::POST,
                None,
            )
            .await
    }

    pub async fn sign(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}/sign", id),
                Method::POST,
                None,
            )
            .await
    }

    pub async fn pdf(&self, id: &str) -> Result<PdfResponse> {
        self.inner
            .request(
                &format!("/api/v1/shipment/ecmr/{}/pdf", id),
                Method::POST,
                None,
            )
            .await
    }
}
